# Release notes

## 1.39 (10 August 2021)

### New/updated features in this release

- Metadata Editor: The special 'Not applicable' and 'Not reported' values are now available in autocomplete suggestions
- Metadata Editor: When filters are being recalculated there is now a loading indicator
- Metadata Editor: S3 links are clickable in the non-curator view
- A new configuration parameter `additionalJwksUris` was introduced to the OAuth provider properties file. This allows the user to specify a list of additional JWKS URIs that can be used for checking token signatures.

### API changes

- Values accepted by the `returnedMetadataFields` parameter (for all endpoints that can return metadata) have changed (previous values still currently work for backwards compatibility). The values are now:
    - `minimal_data` - return only metadata in accordance with the default template
    - `extended_data_included` (previously: `template`) - return metadata in accordance with the applied template
    - `original_data_included` (previously: `all`) - return all metadata attributes
- Endpoints returning dictionary (controlled vocabulary) terms can now also return the identifier for that dictionary term (by supplying the value `term_id` to the `include` parameter).

### Fixed issues

- Metadata Editor: After importing new samples from a spreadsheet the values are available to replace in the bulk replace dialog right away without needing to refresh the page
- Metadata Editor: In the non-curator view it is now possible to filter samples by values with the type `date`.
- The API behaviour for xrefsets endpoints was improved:
    - The endpoints `POST /xrefsets` and `DELETE ​/xrefsets​/{id}` returns `403 “Not enough permissions to work with the xrefset”` if a user is not a member of the Curator Group.
    - The endpoint `GET /xrefsets/{id}/metadata` returns `400` Bad request if a user passes invalid object’s ID (anything apart from Genestack accession).

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>

## 1.38 (28 June 2021)

### New/updated features in this release

- The Metadata Editor has several new features:
    - Absence of values is indicated with a specific `No value` in italics
    - Columns are resizable
    - We have introduced special `Not applicable` and `Not reported` values that are always valid regardless of template rules
    - Autocomplete suggestions for `Not applicable` and `Not reported` values have been added to the bulk replace dialog
    - The bulk replacement dialog will now reveal full values when you hover over them
    - Removed explicit manual pagination from metadata table. Users can scroll samples table infinitely

### Fixed issues

- Fixed an issue when a user without curator permissions has no way to explore an applied template from the Metadata Editor
- Fixed an issue with removing filters when no samples match them
- A bug in omics/expression/data after deleting a xref mapping file was fixed

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>

## 1.37 (25 May 2021)

### New/updated features in this release

- The Users and Permissions page has been updated with a new look and feel. You can use the contextual menu to deactivate/activate particular users
- Error logs/code/other text have improved security
- The URL of the “Groups” page has changed to `/frontend/endpoint/application/run/genestack/web/groups`
- Instances with Arvados integration: Arvados hosts are shown in user profiles, and can be clicked on to open Arvados Workbench
- Instances with Arvados integration: Arvados error messages have been improved

### Fixed issues

- Instances with Arvados integration: A problem with importing studies via the Arvados importer GUI has been fixed.

### Known issues

- The study browser may continue to show and search facet values from objects that have been unlinked from studies
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>

## 1.36 (27 April 2021)

### New/updated features in this release

- Users have the option to expand the Bulk Replace dialogue so the values to be replaced are not truncated
- Swagger help page descriptions have been improved

### Fixed issues

- Swagger descriptions of some endpoint page limits have been corrected
- Sorting using the cursor on omics endpoints now works correctly

### Known issues

- The study browser may continue to show and search facet values from objects that have been unlinked from studies
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>

## 1.35 (30 March 2021)

### New/updated features in this release

- All user accounts will now appear in the users list, including technical superuser accounts
- The organization name will no longer be displayed in the interfaces
- The option to sign in with Okta is now available
- Several pages interfaces have been moved to a new framework

### API changes

- Swagger help page descriptions have been improved

### Fixed issues

- Filters now work for multi-values in the metadata editor
- Bulk-replace operations in the metadata editor now occur in parallel (replacing `A,B` with `A->B` and `B->C` will now result in `B,C` not `C,C`)
- Corrected error messages for permissions related failures
- Gene dictionaries now initialise correctly
- `GET omics/samples` no longer fails with the error `incorrect cursor format`
- Cursor pagination works correctly in omics/samples with no filters

### Known issues

- It is not possible to import public studies since the superloader module has been removed. This functionality will be considered in the future
- The study browser may continue to show and search facet values from objects that have been unlinked from studies
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>

## 1.34 (5 March 2021)

### New/updated features in this release

- The URL for the Study Browser has been changed
- Imported templates and dictionaries are shared with everyone by default
- The default template can be edited - **this change requires the default template be selected again when `1.34` is first deployed**.
- The `import_ODM_data` script now returns the jobExecutionIdentifier to users so imports can be monitors
- Organization properties have been moved to a configuration file

### API changes

- Depreciated API endpoints have been removed

### Fixed issues

- Fixed an error with xref mapping
- Fixed an issue deleting studies via the `studyCurator` endpoints
- Fixed an issue causing invalid sample filter counts when there are invalid field values
- Improved error descriptions for the helper script `create_users.py`

### Known issues

- It is not possible to import public studies since the superloader module has been removed. This functionality will be considered in the future
- The study browser may continue to show and search facet values from objects that have been unlinked from studies
- Filters in the samples table in Metadata Editor do not work for multi-value fields
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>

## 1.33 (27 Jan 2021)

### New/updated features in this release

- **NEW METADATA OBJECT TYPES**: Libraries and Preparations have been added as additional types of objects which users can import, link, edit and query. These optional single sources of truth contain information about how samples for transcriptomics and proteomics studies respectively have been prepared, and are supported with the usual suite of validation and curation tools in ODM
- **NEW ENDPOINTS**: `/libraries` and `/preparations` (which can be found under `libraryUser/Curator` and `preparationUser/Curator` in swagger) have been added to support importing, linking, editing and querying libraries and preparations
- Gene-transcript mapping has been generalised to cross-reference mapping and endpoints have been renamed `/xrefsets`, which can be found under the **reference-data** section in swagger. These endpoints have also been extended for more flexible functionality
- There is a new user permission `CONFIGURE_FACETS`. Users with this permission can configure the facets that are displayed in the Study Browser
- The module for importing public studies has been removed

### Fixed issues

- Fixed some issues and issue help text for the /xrefsets (was gene-transcript mapping) endpoints

### Known issues

- NEW: It is not possible to import public studies since the superloader module has been removed. This functionality will be considered in the future
- The study browser may continue to show and search facet values from objects that have been unlinked from studies
- Filters in the samples table in Metadata Editor do not work for multi-value fields
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>

## 1.32 (30 Dec 2020)

### New/updated features in this release

- The group roles have been simplified to just group administrator and sharing member
- The auto-complete dropdown no longer displays `searching...`, it shows auto-complete matches or an explicit `search for <input value>` option instead
- Starting ODM for the first time or when no data is currently available now displays a more user-friendly page
- Facet value sorting in the Study Browser has been improved to prevent ordering switching too much when selecting multiple data types
- Adding new files to a study now automatically shares them with the same permissions as the study
- New files added to a study automatically have the study template applied to them
- The Python client (used by some helper scripts) is now based on Python 3

### Fixed issues

- Additional improvements have been made to asynchronous data loading services (ETL), in particular when loading large expression files
- Improved error messages when attempting to link objects with no matching Sample Source IDs in the samples table
- Performance in the Study Browser has been improved when there are lots of different facet values
- The share studies option now forces selection from existing groups rather than free text
- Copying dictionary names in the Template Editor now works correctly
- Search text in the Groups application is preserved when making changes to a user’s role
- Fixed an issue exporting expression data that came from an S3 source

### API changes

- Added a new parameter `returnedMetadataFields` to integration endpoints. This parameter controls the amount of metadata that is returned from integration queries. By default, curator endpoints return all available metadata and user endpoints return metadata only for template fields.

### Known issues

- NEW: The study browser may continue to show and search facet values from objects that have been unlinked from studies
- There are currently some issues loading data from GEO
- Filters in the samples table in Metadata Editor do not work for multi-value fields
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>

## 1.31 (11 Nov 2020)

### New/updated features in this release

- Added a user permission `MANAGE_ORGANIZATION` which can be toggled in the Users and Organization application. This replaces the previous (fixed) organization administrator role.
- API endpoints for linking (`POST /integration/link/*`) now better handle situations where a user is importing data for some samples but not all.

### Fixed issues

- Several improvements have been made to asynchronous data loading services (ETL)

### Known issues

- NEW: There are currently some issues loading data from GEO
- Filters in the samples table in Metadata Editor do not work for multi-value fields
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>

## 1.30 (16 Oct 2020)

### New/updated features in this release

- Variant API queries can now be done against multiple variants
- A template can now be chosen as the default template through the GUI
- Pagination of query results for the `omics/*/data` endpoints now uses the cursor parameter instead of `pageoffset`. Results returned include a `cursor` tag. Repeat your query and supply this tag to the cursor parameter to obtain the next page of results and a new cursor tag. When no more results are available the return will be empty.
- `Pageoffset` pagination is now depreciated for `omics/*/data` endpoints

### Fixed issues

- Fixed an issue preventing non-curator users who had manage templates permissions from exploring templates via the Metadata Editor

### Known issues

- Filters in the samples table in Metadata Editor do not work for multi-value fields
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>
