# Release notes

## 1.49

### New/updated features in this release

- The ability to load files to ODM from a local machine with installed ODM via NFS is added. To load files to ODM, users should use Job API endpoints for async loading, e.g., `POST /import/expression` with `"source": "LOCAL"`. Loading mapping files from a local machine via `POST /xrefsets` is not supported.
- The `main()` function is added to the `import_ODM_data.py` script.
- Dictionaries without any term can be loaded to ODM. And if such a dictionary is assigned to an attribute in the template, this attribute is treated as a free text attribute in Metadata Editor. Any text that the user enters is valid.

### Fixed issues

- Fixed: Sometimes, a wrong sample count is displayed in the Study Browser after importing a study to ODM.

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies.

<hr/>

## 1.48

### New/updated features in this release

- API endpoints for working with user accounts have been introduced, all endpoints require the "Manage organisation" permission (user should be Admin in ODM):
    - `GET /groups`: the endpoint returns a list of non-deleted user groups in which the user is a member of the group. Users can be filtered by their attributes, e.g. by `displayName`;
    - User accounts can be created and updated using `POST /users` endpoint;
    - Attributes of a specific user account can be updated by `PATCH /users/{id}` endpoint.
- Improvements in the Metadata Editor:
    - Attribute description and its data type are shown in the attribute menu and the bulk replace dialogue;
    - Share-study dialogue was redesigned. When a user shares a study with a new group, the user can see a list of the groups with which this study has already been shared.
- A new permission “Manage groups” was implemented, it allows to retrieve all user groups.
- Improvements in API:
    - Performance is improved for the cases of a significant amount of patch requests (with disabled metadata versioning on the instance);
    - Optimization of omics query results in faster queries, and the system stability (less memory per processor is used, etc.).
- The code quality and maintainability of the import_ODM_data.py python script for data loading to ODM have been improved:
    - All argument keys use dashes consistently, e.g. `--mapping-file`. Arguments with underscores are still supported, e.g. `--mapping_file`;
    - User can load large files with preparations and libraries (>10 000 rows) via the `Import_ODM_data.py` script since the script uses async API for data loading.
- The documentation covering `update_templates.py` script has been updated.

### Fixed issues

- Fixed: querying signal runs via `GET /expression/runs/by/group/{id}` returns incorrect "total" in pagination info.

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies.

<hr/>

## 1.47

### New/updated features in this release

- The "Description" field is added for attributes in Template Editor. The field can be used to store a brief explanation of what each attribute was intended for, examples of values, etc. The maximum length of the description is 500 characters.
- Attribute descriptions, which provide a brief explanation of what each attribute was intended for, examples of values, etc., are added in Metadata Editor. A description of each attribute can be edited in the study template.
- The "Description" attribute is supported in the python script for loading templates.
- Curate libraries and preparations metadata using the curation script.

### Fixed issues

- Fixed: dates are displayed in milliseconds in the view mode of Metadata (e.g., `"1638448612812"` instead of `"December 2, 2021"`).
- Fixed: when a new template is created using the Duplicate function of the Template Editor, the page does not indicate whether the new template has been generated.

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies.

<hr/>

## 1.46

### New/updated features in this release

- When the user switches between editing and viewing modes in the Metadata Editor, the selected filters are saved (there is no need to choose and apply them again).
- Visual distinction highlight of template vs. non-template attributes is added to all tabs of the Metadata Editor.
- More effortless study curation when changing the applied template or applying a new template that causes a change in the attribute data types. Previously: if e.g., the `age` field is a `string` attribute, it is changed to become an `integer` attribute in a new applied template. This attribute's previously valid values were marked as `invalid`, so the user had to edit each value.<br/>
    Currently: the user can call a mass conversion of the value types according to the applied template.<br/>
    The following type conversions are supported: `Text` -> `Integer`, `Text` -> `Decimal`.
- The user confirmation request was added to the Template Editor for the attribute removal from the template. It avoids accidental removal of the attribute or changing the validation rules used in studies.
- MySQL version has been upgraded to `8.0.28` to avoid vulnerabilities in the previous version.
- ClickHouse is updated from `21.8` to `21.12`. The ClickHouse logs are more straightforward. Investigating ClickHouse problems has become easier.

### Fixed issues

- Fixed: the application returns an error when adding a new facet in the "Configure facets" application with underscores in the name (e.g., `Study_Type` instead of `StudyType`), and saving the changes.
- Fixed: an incorrect number is displayed in the counter on the Samples tab in Metadata Editor after applying filters.
- Fixed: switching from the edit mode to the view mode causes an error "Page unresponsive" while working with a study with more than 10k+ samples in Metadata Editor.
- Fixed: The filters in the Metadata Editor were not updated while the user was editing the sample attributes. E.g., the user filters samples by `Disease = Asthma`; updates the value of one of the attributes from `Asthma` to `Cancer` and does not publish changes; then the sample is continued to be returned in filters by `Disease = Asthma`.
- Fixed: the user needs to refresh the page in the Metadata Editor to publish the changes.
- Fixed: The user receives an error message when reassigning an attribute to a column with integer values on the Samples tab in Metadata Editor.
- Fixed: Obsolete dictionaries were not highlighted in red in the Template Editor application; the user could have mistakenly selected an outdated dictionary for an attribute.

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies.

<hr/>

## 1.45

### New/updated features in this release

- Solr `8.11.1` version is updated to resolve the `log4j` vulnerability.
- Full-text and facet search for studies via API endpoint `POST /integration/fulltext/search/studies`. The endpoint allows searching for studies by study metadata and all linked objects' metadata (e.g., samples, libraries, preparations, omics metadata). Examples:

    a) find all studies where the term "cancer" occurs in the data, e.g., in samples or the study description;

    b) find all studies by a specific facet `"Disease" = "Asthma"` (the endpoint performs search only across facets defined in the "Configure facets "application).

    The endpoint returns:

    - a list of studies with the metadata summary;
    - a list of facet objects with counts (the same list of facets shown in the "Study Browser" application).

- Asynchronous loading of files with "preparation" information via API endpoint `POST /import/preparations` to upload files with more than 10k rows.
- The performance of filters in Metadata Editor in edit mode was improved (1.5 times faster processing on the Genestack test environment). The exact time difference depends on the instance settings and the connection provider.
- Ability to copy values and reassign attributes for both template and non-template attributes in Metadata Editor.
- The `import_ODM_data` python script now supports versions of Omics data files. Thus, the user can upload a new omics data file to an existing study using the "import_ODM_data" python script as the next version of the previously uploaded file. The solution has limitations described in the python script documentation.
- Synchronous uploading performance of samples via the API endpoint `POST /samples` was improved (2.7 times faster processing on the Genestack test environment). The exact time difference depends on the instance settings and the connection provider. Now optional "Warning" attribute with the results of sample validation for the applied template is not returned in the response body.
- The deprecated endpoint for loading cross-reference (xref) mapping files `POST ​/import​/transcript-mapping` method has been removed; the endpoint call returns an error. For cross-reference mapping files loading user can use `POST /xrefsets`.
- The "View Logs" button is removed from the Export page.

### Fixed issues

- **Breaking changes!** The method `get_parents_by_study` was renamed to `get_groups_by_study` in Python and R SDK. If your application uses the methods `get_parents_by_study`, rename all the method usages to `get_groups_by_study`. If your application does not use R or Python SDK, no changes are required.
- If the mapping file contains duplicates, the endpoint loads only unique rows. A warning regarding duplicates is returned in the response body, e.g., "The loaded file includes three duplicate rows that were skipped". Previously, when the user loaded mapping files with duplicate rows via `POST /xrefsets`, the endpoint stopped the file loading on the first duplicate and did not return the warning message. The user did not know whether the entire file was loaded or not.
- When a user applies another template to a study in Metadata Editor, the new validation rules are immediately applied. It allows the user to see the actual data, the actual set of required attributes, and the actual status of the data validation. Previously, the user needed to refresh the page to enable it.
- Fixed: when a user adds and removes the multi-value field in Study Metadata, the error is returned.
- R SDK documentation is improved:
    - An example "include "parameter value is added to the description.
    - Clients receive an exception when passing unknown parameters in functions.
- Fixed: API endpoints for removing links between objects return code `204` if the specified object of a wrong type or there was no link between objects. The endpoints return `404` with a message corresponding to the case. Affected `DELETE` endpoints:
    - `/integration/link/sample/{sourceId}/to/study/{targetId}`
    - `/integration/link/sample/group/{sourceId}/to/study/{targetId}`
    - `/integration/link/library/{sourceId}/to/sample/{targetId}`
    - `/integration/link/library/group/{sourceId}/to/sample/group/{targetId}`
    - `/integration/link/preparation/{sourceId}/to/sample/{targetId}`
    - `/integration/link/preparation/group/{sourceId}/to/sample/group/{targetId}`
- Fixed: the swagger page does not become unresponsive when the user calls the endpoint `GET /sampleCurator/samples` without specified `pageLimit`, and there are more than 2000 samples.

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies.

<hr/>

## 1.44

!!!info "Deployment note"

    Versions from `1.42` onwards use a new schema for core ClickHouse tables. The previously released version `1.41` includes migration scripts so deployment of `1.41` and completion of this migration job must have taken place before deploying any version from `1.42` onwards.

### New/updated features in this release

- Filters in the Metadata Editor are now applied faster.
- The Metadata Editor allows scrolling across the table.
- Ontology-aware search for RDF-formatted ontologies: When using the search bar in the Study Browser exact and related synonyms are added to the search query by default (where the user’s query matches a term from an ontology). Users can also extend the results of a query by toggling inclusion of all child (descending) terms.
- Study Browser results summaries now include metadata from these additional fields: `Data Species`, `Species Name`, `Sample Type` and `Disease Status`.
- Users can load libraries data objects asynchronously via a new endpoint `POST /import/libraries`. The endpoint returns a `jobExecId`, and the results can be checked via the Job operations endpoint.

### Fixed issues

- Critical security vulnerability was fixed in `log4j`.
- Fixed exporting omics data where metadata versioning feature is enabled.
- Edit mode in the Metadata Editor no longer causes timeouts with large (50,000) samples tables.
- Reassigning attribute now correctly deletes the source column from view mode as well as from edit mode in the Metadata Editor.
- Metadata Editor now has correct filters descriptions when values are updated.
- Links are recognized and clickable in the view mode of the study metadata tab.
- Users now see the correct facets values for bulk replace after discarding changes.
- The correct number of available samples is shown after filtering.
- Incorrect use of the template was fixed for Bulk replace in Library/Preparation metadata.
- Fixed a problem with unpublished versions after turning on metadata versioning feature.
- Curation script works for Variant, Expression and Flow Cytometry metadata as well as for tabs `Study` and `Samples`
- Loading of `vcf` files with values `nan` was fixed

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies.

<hr/>

## 1.43

!!!note "Deployment note"

    Versions from `1.42` onwards use a new schema for core ClickHouse tables. The previously released version `1.41` includes migration scripts so deployment of `1.41` and completion of this migration job must have taken place before deploying any version from `1.42` onwards.

### New/updated features in this release

- Full text search in the Study Browser now supports ontology expansion - if a search term matches a term from an ontology in RDF format then it is possible to extend the query to include child terms (those with the `subClassOf` property), up to a limit of 30,000 child terms (and synonyms).
- Full text search now automatically includes exact and related synonyms (`hasRelatedSynonym` or `hasExactSynonym` in the RDF)
- There is a new `roundDigits` request parameter for the /streamed-data endpoint. By default `roundDigits = 4`

### Fixed issues

- Metadata Editor: Renaming a study via the contextual menu now creates a new version (if metadata versioning is enabled).
- Users who singed in using SSO (`OAuth2`) now can re-sign in via one click again after being logged out due to security reasons.
- Now it is possible to undo (<kbd>ctrl+z</kbd>) and redo (<kbd>ctrl+shift+z</kbd>) manual changes to metadata.
- The metadata validation summary now correctly reports valid metadata when all metadata is valid.
- `UserEndpoint` APIs return latest draft (staging) versions of signals metadata when versioning feature is disabled, ensuring users get the latest version of the data.
- Swagger endpoints have had some extraneous descriptions removed.
- Swagger description for `GET /omics/expression/streamed-data` has been improved.

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies.

<hr/>

## 1.42

!!!note "Deployment note"

    Versions from `1.42` onwards use a new schema for core ClickHouse tables. The previously released version `1.41` includes migration scripts so deployment of `1.41` and completion of this migration job must have taken place before deploying any version from `1.42` onwards.

### New/updated features in this release

- Metadata versioning is available as a beta feature. To enable it see the administration guide. By default it is not enabled.
- The `POST /task/publish-versions` API endpoint was added to support metadata versioning.

### Fixed issues

- Metadata Editor: Applying a new template is now reflected immediately without needing a page refresh.
- Metadata Editor: Copying/previewing sample accessions now work correctly with pagination.
- A bug with session tokens after a server restart has been fixed.
- Adding samples via `POST /integration​/link​/sample​/group​/…​/to​/study​/…` now correctly shares these samples with the same permissions as the study being added to.
- The Uberon ontology now successfully initiates on instances with HTTP(S) proxies.

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies.

<hr/>

## 1.41

!!!note "Deployment note"

    This release contains a periodic job which automatically migrates core ClickHouse tables (`ExpressionDataByRun`, `ExpressionDataByGene`) to a new schema. It is mandatory to let this job finish before installing future releases (`1.42` etc.) as they will not support the old schema. The job is finished when the backend logs contain the messages: `Migration of data for table {} finished`, where `{}` is `ExpressionDataByRun` and `ExpressionDataByGene`.

### New/updated features in this release

- Minor dashboard text changes
- The Template Editor URL has changed to a Base URL of: `frontend/endpoint/application/run/genestack/web/templates`<br/>
    Redirects are in place for the previous URL.
- Improved use of server resources (disk space in particular)
- Improved performance of omics data queries

### Fixed issues

- Metadata Editor: Fixed an issue filtering by invalid values works for Integer/Date data types. Now users can use date/integer values for filtering even if they are invalid according to the template
- Metadata Editor: Dictionary terms are now correctly highlighted after changing template.
- An Expression data query performance degradation bug has been fixed
- Patch API calls now work on large sample numbers
- Omics queries: Filtering of variants by INFO fields in text format is working now, e.g. `vxQuery = info_VT=INDEL`

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>

## 1.40

!!!note ""

    Customers should update their SSL settings to use a `2048-bit` or greater DH cipher

### New/updated features in this release

- Study Browser: Added a syntax cheatsheet that describes how to use advanced full-text search capabilities
- Metadata Editor: Added a new export button to export all data/metadata of a study
- Export: Exports now include a readme detailing the files contained in the archive
- API endpoints: A `400` http code is now returned when an unknown parameter is supplied to any endpoint

### Fixed issues

- Metadata Editor: Fixed an issue editing fields at the bottom of the window
- Study browser: Fixed saved bookmark icons
- Omics data files are exported with the same extensions that they were imported with

### Known issues

- Metadata Editor: There is no automatic formatting of numeric values, they will be displayed as strings – aligned to left instead of right.
- The study browser may continue to show and search facet values from objects that have been unlinked from studies
- The curation helper script doesn't work for Variant, Expression and Flow Cytometry objects, only for studies and samples

<hr/>
