# Release notes

!!!info "Target users"

    **Curators** - users who add new data into ODM and are responsible for the data harmonization and curation. That includes creating and defining metadata templates, mapping metadata and temples, and data updates.

    **Researchers** - users who access ODM to identify a batch of data suitable for further research and analysis. That includes search, data browsing and export.

    **Advanced** - users who can utilize advanced API functionalities for user management and data management.

    **Admins** - users who manage organization in ODM, its user and groups.

## 1.59 

## 1.58 Open Data Manager 2024 Summer Release: VCF functionality

### New Features

We've completely overhauled the support for variant data in VCF files to enhance performance with larger files.

!!! tip "Previously uploaded VCF data will be automatically migrated to the new output formats only if the source link is still valid and data can be accessible via it."

#### [Curators]

**Support now includes the ability to upload compressed VCF files**.
We have also increased the speed of file uploading and indexing for these large files at the following API endpoints:

* POST `/api/v1/jobs/import/variant`
* POST `/api/v1/as-curator/variants/vcf`

Additionally, the system can now handle up to ten VCF files being uploaded simultaneously.

#### [Researchers]

1. **Enhanced parsing functionality:**
Uploading new VCF files now ensures that the INFO and Genotype fields are correctly split and
parsed based on the data types (String and Numeric) specified in the VCF file header.** This enhancement aids in
constructing more detailed filters for variant identification.
2. **Enhanced vxQuery Parameter:**
The vxQuery parameter has been enhanced to support variant filtering based on `filter`, `id`, and `region` keywords,
along with their logical combinations using `not (!)`, `and (&&)`, and `or (||)` operators, respecting standard
Boolean precedence.

     * ??? abstract "vxQuery Parameter abilities"
        * `filter` corresponds to the `FILTER` column in the original vcf file. Could be either equals or not equals
        to `PASS` : `filter = PASS` or `filter != PASS`.
        * `id` is a string value that corresponds to the `ID` column. Multiple values could be added via a comma.
        Expression `id = rs3214,rg1234` is equivalent to `id = rs3214 or id = rg1234`.
        * `region` corresponds to CHROM and POS columns. Supports querying by the whole chromosome `region = chr1`,
        by interval within a chromosome `region = chr1:4567-9876543` by exact position `region = chr1:456789` or by
        indefinite ranges like `region = x:3456789-*` . Multiple regions could be separated via comma,
        e.g. `region = chr1:34567-98767,chr3` is equivalent to `region = chr1:34567-98767 or region = chr3`.
  
     * ??? danger "Affected endpoints (for both as-curator and as-users) that include vxQuery parameter"  
         * GET `/api/v1/as-curator/omics/expression/data`
         * GET `/api/v1/as-curator/omics/expression/group`
         * GET `/api/v1/as-curator/omics/flow-cytometry/data`  
         * GET `/api/v1/as-curator/omics/flow-cytometry/group`  
         * GET `/api/v1/as-curator/omics/samples`  
         * GET `/api/v1/as-curator/omics/variant/data`  
         * GET `/api/v1/as-curator/omics/variant/group`

3. **Retrieving Original VCF Annotation Data**
When a new VCF file is uploaded, the format and descriptions from the VCF header are stored and available for query.
This allows users to understand available fields for filtering. Annotation details are added under the “metadata”
section in the response for relevant endpoints.
    * ??? danger "Affected endpoints (for both as-curator and as-users)"
        * GET `/api/v1/as-user/variants/group`  
        * GET `/api/v1/as-user/variants/group/by/run/{id}`  
        * GET `/api/v1/as-user/variants/group/{id}`  
        * GET `/api/v1/as-user/omics/variants/group`

4. **Data Export in VCF Format**
    1. Users can now stream data from a specific VCF file group, applying:
        1. Variant filters to export only the required set of variants.
        2. Metadata filter to export only the samples corresponding to the specified metadata attribute values.
    2. Unfiltered requests will export the complete dataset.
    3. Endpoints introduced:
        * GET `/api/v1/as-curator/omics/variant/streamed-data`
        * GET `/api/v1/as-user/omics/variant/streamed-data`

5. **Changed response format for all data types (Variants, Expression, and Flow cytometry)**

    We are implementing updates to our API product to provide a more consistent and efficient experience.
The updates include changes to the response format, the addition of a new parameter, and the deprecation of outdated
features.
The response format changes affect the response format only for the GET endpoints for Tabular
(including Gene Expression), Variant, and Flow Cytometry data that provide output in a JSON-like format.
Streaming GET endpoints that provide output in a tabular format are not affected.
Please review the documentation, as these changes may impact your integrations.

    1. **Common changes for all data types**

           * ??? abstract "Introduced changes to all data types"
               | Introduced change                                   | Changes Description         |
               |:----------------------------------------------------|:----------------------------|
               | New “**itemOrigin**” section in the response format | A new section **“itemOrigin”** is introduced in the response body to store information regarding the origin of the data item: **runSourceId** - run id from the original file, **runId** - Genestack run id generated upon import, **groupId** - id of the original file. |
               ||“**Run Source ID**” is moved from the metadata section to “itemOrigin” and renamed to “runSourceId”. In the OMICS Endpoints (integrationCurator, integrationUser) **sampleFilter**, **libraryFilter** or **preparationFilter** can be used for retrieving an individual run.|
               | New “**runSourceFilter**” parameter                 | A new parameter allows filtering data by the Run IDs from the original files.  |
               | “**runFilter**” parameter takes list of strings     | “**runFilter**” parameter which is taking the auto generated internal id for runs takes list of strings |
               | Deprecated filter option                            | FIltering by “**Run Source ID**” in the metadata filter is not available.     |

           * ??? example "Response example"
    ![RN](1.58/1-58-1.png)

           * ??? danger "Affected endpoints (for both as-curator and as-users)"
               | Endpoint Group       | Endpoints Affected                                                                                                                      |
               |:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|
               | expressionUser       | GET `/api/v1/as-user/expressions` GET `/api/v1/as-user/expressions/{id}`                                                                    |
               | expressionCurator    | GET `/api/v1/as-curator/expressions` GET `/api/v1/as-curator/expressions/{id}`                                                              |
               | variantUser          | GET `/api/v1/as-user/variants` GET `/api/v1/as-user/variants/{id}`                                                                          |
               | variantCurator       | GET `/api/v1/as-curator/variants` GET `/api/v1/as-curator/variants/{id}`                                                                    |
               | flowCytometryUser    | GET `/api/v1/as-user/flow-cytometries` GET `/api/v1/as-user/flow-cytometries/{id}`                                                          |
               | flowCytometryCurator | GET `/api/v1/as-curator/flow-cytometries` GET `/api/v1/as-curator/flow-cytometries/{id}`                                                    |
               | integrationUser      | GET `/api/v1/as-user/omics/expression/data` GET `/api/v1/as-user/omics/variant/data` GET `/api/v1/as-user/omics/flow-cytometry/data`          |
               | integrationCurator   | GET `/api/v1/as-curator/omics/expression/data` GET `/api/v1/as-curator/omics/variant/data` GET `/api/v1/as-curator/omics/flow-cytometry/data` |

    1. **Tabular Data Endpoint Changes**

          * ??? abstract "Introduced changes to Tabular Data"
              | Introduced change                                   | Changes Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
              |:----------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
              | Deprecating old response format | The old response format for tabular data will be deprecated. The **"multi_value"** response format (introduced in July 2023) will be the default option. It allows retrieval of data with one or more features and suits any type of tabular data, including **.gct** files. |
              | Changing field name | The **“gene”** field is renamed to **“feature”** and incorporates all the feature attributes. The **“expression”** field with the measurement value for the **“feature”** was renamed to **“value”** and placed in the new **“value”** section. |

          * ??? example "Response example"
      ![RN](1.58/1-58-2.png)

          * ??? danger "Affected endpoints (for both as-curator and as-users)"
              | Endpoint Group       | Endpoints Affected                                                                                                                      |
              |:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|
              | expressionUser       | GET `/api/v1/as-user/expressions` GET `/api/v1/as-user/expressions/{id}`                                                                    |
              | expressionCurator    | GET `/api/v1/as-curator/expressions` GET `/api/v1/as-curator/expressions/{id}`                                                              |
              | integrationUser      | GET `/api/v1/as-user/omics/expression/data`  |
              | integrationCurator   | GET `/api/v1/as-curator/omics/expression/data` |

    2. **Variants Endpoint Changes**

          * ??? abstract "Introduced changes to Variants"
              | Introduced change                                   | Changes Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
              |:----------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
              | Introducing **“variant”** section | A new section **“variant”** contains all fields related to a specific variation from the original VCF file: CHROM, POS, ID, REF, ALT, QUAL, FILTER, INFO. |

          * ??? example "Response example"
      ![RN](1.58/1-58-3.png)

          * ??? danger "Affected endpoints (for both as-curator and as-users)"
              | Endpoint Group       | Endpoints Affected                                                                                                                      |
              |:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|
              | variantUser          | GET `/api/v1/as-user/variants` GET `/api/v1/as-user/variants/{id}`                                                                          |
              | variantCurator       | GET `/api/v1/as-curator/variants` GET `/api/v1/as-curator/variants/{id}`                                                                    |
              | integrationUser      | GET `/api/v1/as-user/omics/variant/data`   |
              | integrationCurator   | GET `/api/v1/as-curator/omics/variant/data`|

    3. **Flow Cytometry Endpoint Changes**

          * ??? abstract "Introduced changes to Flow Cytometry"
              | Introduced change                                   | Changes Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
              |:----------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
              | Introducing **“feature”** and **“value”** sections | Flow-cytometry data response now has section **“feature”** including **"readoutType"**, **"cellPopulation"** and **"marker"** fields. **“Value”** section contains the expression field which is renamed to **“value”**.|
              | Changing field name             | The old **“expression”** field with the measurement for the **“feature”** was renamed to **“value”** and placed into the new “value” section. |

          * ??? example "Response example"
      ![RN](1.58/1-58-4.png)

          * ??? danger "Affected endpoints (for both as-curator and as-users)"
              | Endpoint Group       | Endpoints Affected                                                                                                                      |
              |:---------------------|:----------------------------------------------------------------------------------------------------------------------------------------|
              | flowCytometryUser    | GET `/api/v1/as-user/flow-cytometries` GET `/api/v1/as-user/flow-cytometries/{id}`                                                          |
              | flowCytometryCurator | GET `/api/v1/as-curator/flow-cytometries` GET `/api/v1/as-curator/flow-cytometries/{id}`                                                    |
              | integrationUser      | GET `/api/v1/as-user/omics/flow-cytometry/data`          |
              | integrationCurator   | GET `/api/v1/as-curator/omics/flow-cytometry/data` |

#### [All Users]

1. **User Guide Update**: A revised user guide is now available, featuring a series of new quick start guide articles.
    * [Quick Start Guide](../user-guide/quick-start/index.md) - introduction to User Interface and API documentation.
      * [Advanced User Interface Guide](../user-guide/index.md)
      * !!! note "We are working on making our guide better. Please reach us in case of any inconsistency."

2. **Lipidomics Data Class**: A new **Lipidomics** data class is available for selection during the upload of tabular data files.
This new class is integrated within the metadata editor and study browser.
    ![RN](1.58/1-58-5.png)

### Fixed Issues

#### [Curators]

1. **Data Curation in Metadata Editor**: It's now possible to delete a sub-attribute value, which will remove the
entire compound attribute value.
2. **Data Deletion**: Fixed an issue where the deletion of a related tabular data file could make an entire study
unavailable.

#### [Researchers]

1. **Data Exploration**: Attribute count will now recalculate upon more then one attribute selection for plot creation.
2. **Data Export**: An error message will now display when attempting to export a data file that has been deleted
from the original AWS S3 storage, prompting users to contact the study owner.

## 1.57 Open Data Manager 2024 Spring Release

### New Features

**[Curators]** Samples Import via the User Interface:

* Support for importing up to 10 000 samples.
* The preview pop-up now allows checking for 1 000 samples before uploading.
  ![RN](1.57/1-57-1.png)

An example of a Sample uploading preview pop-up. The file contains 2713 samples, while preview allows browsing the
first 1000 rows of the file prior to uploading.

**[Curators]** Libraries and Preparation Import via the User Interface (previously available only in the API):

* Added a "+More" button to select the uploading of Libraries and Preparation.
* Increased import limit for a Libraries and Preparations files to 128 MB.
  ![RN](1.57/1-57-2.png)

Libraries/Preparations file requirements: all Sample IDs mentioned in the file must match Sample IDs previously
uploaded in the Samples tab for the selected Study

**[Researchers]** Data Exploration for Samples metadata:

* A new ‘Explore’ button on the Study page launches sample data visualization.
* Explore view enriched with:
    * Sample attribute selection capabilities. Each attribute is associated with the total number of unique values
  and the nature of the attribute: "string" for categorical and "numeric" for continuous attributes.
    ![RN](1.57/1-57-3.png)
    * Various plot types, including Bar plot, Histogram, Scatter plot, Category plot, and Box plot, are available
  based on the selected attribute type.
    ![RN](1.57/1-57-4.png)
    * Automatic data filtering based on validation rules (removes empty and invalid values from visualization).
    * Export function for plots available in PNG or SVG format.
    ![RN](1.57/1-57-5.png)
    * Instructions for Working with Data Exploration. Included comprehensive guidelines for utilizing Data Exploration
  features effectively.
    ![RN](1.57/1-57-6.png)

### Fixed Issues

**Improved Dictionaries:**

* **[Curators/Researchers]** The original dictionary details info appears while hovering over values in the Sample
Data Tab.
  ![RN](1.57/1-57-7.png)
* **[Curators]** Fixed the ability to replace an old dictionary with a new one via the script.

## 1.56 Open Data Manager 2024 February Release

### New Features

* **[Advanced] Usage Tracking**: this feature allows to collect metrics and statistics organised in dashboards on Genestack side. Information for particular instance can be requested from Genestack. Currently available metrics are:
    * Number of users with “Activated” status.
    * Number of API calls broken down by API labels.
    * Average length of UI sessions.
    * Number of Studies.
    * Number of metadata objects and groups for Samples, Libraries, and Preparations.
    * Number of data objects broken down by Data Classes.
* **Data Class labels** per Study are reintroduces from Study Browser view.
* **[Curators] Data upload via GUI**: gct/generic data/vcf/facs data can be associated with any sample metadata attribute from template (except the key genestack:accession) when uploading data via Metadata Editor. This prevents users from need to edit original data to meet the rules of ODM data model.
* **Templates**: Ability to rename template when duplicating it.

### Fixed Major Issues

* **GUI Issues**:
    * Fixed error message when linking data to Sample metadata attribute with empty values in Study.

### Other Changes

* **Tasks Page**: redesigned Task Manager is available from GUI.
* **Python SDK, R SDK** are not published in the release.

## 1.55 Open Data Manager 2023 Winter Release

### New Features

* **[Curators, Advanced, Admin] Manage Data**
    * **New Swagger Page**: Introducing new endpoints for managing detached objects and deleting objects in ODM.
        * `GET/detached-objects`: Identifies detached data objects (Samples, Libraries, Preparations, Tabular Data, VCF, and Flow Cytometry data) that are not linked to other objects, aiding in clean-up after incomplete data uploads or deletions.
        * `DELETE/data`: Facilitates cascade deletion of selected objects from ODM. For instance, deleting a Study also removes all linked objects (Samples, Libraries, Tabular Data, VCF, etc.). Access restricted to users with `MANAGE_ORGANIZATION` and `ACCESS_ALL_DATA` permissions.
* **[Curators] Data Uploading Enhancements in GUI**
    * **Enhanced Tabular Data Handling**: Supports specifying a separator character in file column headers for distinguishing between sample names and measurement types.
    * **Advanced Options**:
        * **Skip Zeroes (Sparse Data Matrix)**: Option to ignore cells with '0' values to optimize performance for datasets common in single-cell technologies.
    * **Pre-validation**: All data files uploaded from the local computer undergo a formatting compliance check before uploading.
    * **File Attachment Updates**: If an attached file already exists in ODM, the system offers options to overwrite or rename the file.

### Fixed Major Issues

* **GUI Issues**:

    * [Curators] Fixed the issue where Bulk Replace doesn't replace some empty values in the Metadata Editor.
    * [Curators] Resolved incorrect behavior when copying/reassigning string values to integer/decimal types.
    * [Curators] Addressed the issue preventing the import of Sample Metadata to a Study with existing attached files or failed tabular data uploads.
    * [Users] Corrected the issue with incorrect data file names in export archives.

* **Scripts and API Issues**:
    * [Curators, Advanced] Fixed the error occurring during the simultaneous upload of multiple files with over 2000 samples.
    * [Curators, Advanced] Resolved the issue where the curation script does not publish changes.
    * [Curators, Advanced] Aligned the behavior of the curation script with GUI functionality, particularly for the "Re-assign" feature.
    * [Advanced] Addressed issues with streaming data endpoints, including excessive log entries, incomplete results under high demand, and malfunctioning with custom attribute-linked Tabular Data.
    * [Advanced] Fixed errors related to long IP addresses and unexpected errors with Access Token requests.

### Other Changes

* **Data Archive Removal**: All export data archives previously generated in the ODM GUI will be deleted. New export archives can be generated on demand.
* **Custom SSE-KMS Key**: Introduced the ability to specify a custom SSE-KMS key for data uploading in the ODM GUI.

<hr/>

## 1.54 Open Data Manager 2023 Summer Release

ODM – renamed to **Open** Data Manager.

### [All users] Expanded Data Capabilities

ODM now supports TSV-formatted data frames. This expansion beyond GCT 1.2, VCF, and FACS formats enables the upload of diverse data types such as proteomics, metabolomics, epigenomics, and analysis results. This requires adherence to the data frame specifications provided in the [Usage](https://odm-user-guide.readthedocs.io/en/latest/doc-odm-user-guide/supported-formats.html#tabular-data) section.

Simply put, your data file can now accommodate multiple columns defining the measured feature and more than one measurement column per sample.

Example of Mass Spectrometry data: Feature columns can encompass: `Gene Name`, `Protein Name`, `Peptide sequence`, `PTM site`, `M/Z ratio`, `Retention time`, etc. Measurements per sample may include: `Signal intensity`, `Signal quality`, etc.

Data upload is possible via the GUI/API, and querying can be executed via the API (see details below).

### [Curators] Enhanced Data Uploading

The revised data uploading function is available as a **BETA** version in both the API and GUI. This feature will be improved in upcoming releases.

### [Advanced] API Improvements

At present, uploading of TSV-formatted data is executed through Expression endpoints, with future releases introducing new endpoints for this operation.

Utilize the following endpoints to upload TSV or bulk transcriptomics data in GCT 1.2:

`post/expression/gct`

`post/import/expression`

We've added two new parameters for these endpoints (detailed descriptions available on corresponding Swagger pages):

* `numberOfFeatureAttributes` - to specify the number of feature columns in the file.
* `dataClass` - to denote the data type (e.g., proteomics, metabolomics, etc.).

### [Curators] GUI Adjustments

The `Create new study` button has been relocated to the center of the screen.

![Dashboard](1.54/1.54-dashboard.png)

The `Add Data (BETA)` button has been incorporated into the new Data tab (further details about the tab below).

<figure markdown>
  ![Add data button](1.54/1.54-add-data-button.png){width=360}
</figure>

You now have the ability to import TSV, GCT, VCF, or FACS data from your local computer or via a direct link.

<figure markdown>
  ![Define your data dialog](1.54/1.54-data-dialog.png){width=480}
</figure>

Moreover, this button can also be used to attach any file to the study.

<figure markdown>
  ![Attach file dialog](1.54/1.54-data-dialog-attach-file.png){width=480}
</figure>

Please refer to the [Usage](https://odm-user-guide.readthedocs.io/en/latest/doc-odm-user-guide/import-samples-spreadsheet.html#import-data-beta) section for more information on these functionalities.

The Metadata Editor now includes an option to add new custom fields for Study and Data metadata.

<figure markdown>
  ![Add another attribute button](1.54/1.54-add-another-attribute.png){width=280}
</figure>

### [Researchers] Enhanced Data Browsing and Export

The `Signal Type` filter has been replaced with the Data Class filter, which represents all available data types across studies.

<figure markdown>
  ![Data Classes](1.54/1.54-data-classes.png){width=320}
</figure>

The `Signal Type` column has been removed from the main study browser panel.

The Expression/Variant/Flow Cytometry tabs have been removed from the Metadata Editor page.

A new `Data` tab has been added to the Metadata Editor page. This tab displays a list of all data files uploaded for the study, organized by their respective `Data Class` parameter (selected during file upload) on the left panel.

<figure markdown>
  ![Files panel](1.54/1.54-files-panel.png){width=320}
</figure>

Clicking on a file enables metadata editing [Curators only] and browsing.

The metadata now includes four mandatory read-only fields that are auto-generated:

* Data Class - the data type selected during file upload.
* Features (string) - string feature column names present in the file (e.g., `Gene Name`).
* Features (numeric) - numeric feature column names present in the file (e.g., `M/Z ratio`).
* Value (numeric) - measurement types if more than one measurement is available per sample/library/preparation (e.g., `Fold Change`, `p-value`).
    <br/>![Data settings](1.54/1.54-data-settings.png){width=300}

These fields are implemented to make the content of these files visible and searchable for data science users. We strongly advise against editing these fields in the template editor, as it could make these files inaccessible. If data is uploaded in the GCT format, `Features (string)`, `Features (numeric)`, and `Value (numeric)` fields are left empty (due to the standard structure of GCT files).

### [Advanced] API Retrieval Improvements

The uploaded data can be searched and extracted using the API endpoint. Similar to the uploading process, data retrieval is performed through Expression parameters of Omics queries (Integration group), with future releases set to introduce new endpoints for this operation.

<figure markdown>
  ![Swagger endpoints](1.54/1.54-swagger.png){width=640}
</figure>

Each endpoint of Omics queries now has two updated parameters:

* In the `responseFormat` parameter, use the `multi_values` value to extract data in the new format. If your data has multiple features/measurements, they will be extracted in full. If the parameter is unspecified, the output returns only the first feature column and the first measurement per sample. This modification was introduced to maintain backward compatibility for applications built for GCT data extraction.
* `exQuery` has been enhanced and now supports flexible searches across files of any content. Search parameters in `exQuery` include options for feature and measurement filtering:
    * To search by feature (feature keyword), specify the required column name, e.g., `feature.Genes="ZNF814"`.
    * To search by measurement type (value keyword), specify the required column name, e.g., `value.intensity > A`.

New search capabilities include exact matches, multiple matches, range filtering, substring search, search by missing values, and more. Please refer to the corresponding endpoint descriptions in Swagger for details.

### Known Issue

If more than one file with the same name and extension is imported or attached via `Add Data (BETA)` button, then `Export` of the entire Study in one archive does not work.

Do not import/attach files with exactly the same name (and extension) to avoid this issue.
In case, two files with the same names were uploaded, you need to download them one by one.

<figure markdown>
  ![Export data](1.54/1.54-export.png){width=420}
</figure>

<hr/>

## 1.52

### GUI features

#### [Curators] Revoking access to a study

Added an ability to unshare a study (revoke access to a study) from a certain user group in Metadata Editor. The feature is available only to the study owner.

<figure markdown>
  ![Revoking access](1.52/1.52-revoking-access.png){width=400}
</figure>

#### [Curators] Removing a multivalue while editing samples in Metadata Editor

Users can manually remove one of the values in a row with multiple values while editing a metadata sample in Metadata Editor. Previously, multi-valued "No Value" rows were automatically removed only after a new version of the sample metadata was published.

<figure markdown>
  ![Remove value](1.52/1.52-remove-value.png){width=320}
</figure>

#### Fixed issues

[Curators].[Metadata Editor] Sometimes sample attributes (with attached dictionaries) with empty values are marked in Metadata Editor as invalid.

[Curators].[Metadata Editor] When multi-value rows in the composite sample attributes have no value, they are merged and not shown in Metadata Editor. In the example, there are "Compound Treatment/Compound", "Compound Treatment/Dose", and "Compound Treatment/Unit"; it is impossible to determine which "Dose" values correspond to a proper "Compound".

<figure markdown>
  ![Multi-values. Before fixing the issue](1.52/1.52-multivalues-before-fix.png){width=420}
  <figcaption>Before fixing the issue</figcaption>
</figure>

<figure markdown>
  ![Multi-values. After fixing the issue](1.52/1.52-multivalues-after-fix.png){width=420}
  <figcaption>After fixing the issue</figcaption>
</figure>

[Curators].[Metadata Editor] The default dictionary "Cell type" can not be used during the validation of the samples in Metadata Editor. Also, this dictionary cannot be reloaded into ODM.

[Curators].[Metadata Editor] Incorrect sample count is shown on the bulk replace attribute editor dialogue.

[Curators].[Template Editor] The exported json file with template attributes contains only the fields visible in the Template Editor. We have removed the outdated "isSingle" attribute from the file, which could lead to errors when reloading the template to ODM.

### API Changes

#### Fixed issues

[Advanced users] An issue with more than one AND operator in the omics metadata filters.

<hr/>

## 1.51

### New/updated features in this release

**API. Updating the feature of the data upload via NFS** [Curators, Advanced].

??? example "Click here to expand"

    Сurators can import study files from a locally mounted Network File Storage (NFS) to ODM with "write" OR "read-only" access permission. There is no need to configure the “write” permission for the files, as it was before.

**Authentication in API via access token of an identity provider** [Curators, Advanced].

??? example "Click here to expand"

    ODM API and the "import_ODM_data.py" script allow authentication with a Genestack API token OR with an access token of another identity provider. To specify a custom access token, use the "Authorization" header; to set the Genestack API token, use the "Genestack-Api-Token" header.

    Note: access token takes precedence, meaning that if both tokens are supplied, the access token will be used for processing the request. The expansion fully supports Azure Active Directory access tokens. For a different custom Identity Provider, please test ODM API beforehand.

**Linking samples and omics data by any custom key via API** [Curators, Advanced].

??? example "Click here to expand"

    Previously: samples from metadata and omics data files could be linked only using the "Sample Source Id" attribute as a reference. Now curators can choose any sample metadata attribute (column) as a reference to link samples.

    Notes: only template attributes can be used for the linkage; if the value of the sample attribute for the linking changes, the relationship between samples and omics data is not updated automatically. More details about this feature you can find in the swagger, the endpoint `POST/integration/link/expression/group/{sourceId}/to/sample/group/{targetId}`, `linkingAttribute` parameter.

### Fixed issues in the metadata curation process

[Curators] Fixed: a template attribute with invalid metadata is marked as "non-template" when a curator is editing sample metadata in Metadata Editor via the bulk replace.

[Curators] Fixed: a formatting error appears for the study metadata view-only mode when the text in an attribute is not transferred to the next row.

[Curators] Fixed: other minor bugs in Template Editor and Metadata Editor.

### Fixed issues in API

[Advanced] Fixed: the API endpoint GET/omics/expression/streamed-data does not return a proper error when a user specifies a wrong accession of the signal group in the groupAccession parameter. Now the endpoint returns an error 404 Not Found with the message `"Group 'N' could not be found"`.

<hr/>

## 1.50

### New/updated features in this release

**SCIM API for managing users and groups in ODM via Azure Active Directory** [Admins].

??? example "Click here to expand"

    We have added SCIM API endpoints that allow the organization admin to manage users and user groups in ODM via Azure Active Directory. You can find all information about these endpoints and their limitations in Swagger in the "Manage Organization" section. Active Directory and ODM integration via SCIM endpoints instruction is located in the admin guides.

    Since the API endpoints have been developed according to the SCIM 2.0 specification, you can use them with any identity provider that supports working with SCIM. But since each identity provider has custom specifications, we recommend the solution pre-testing.

    We recommend calling user and group synchronization in Active Directory on behalf of the user with the "Manage Groups" permission so that the user can see and edit all groups on the instance, regardless of whether the user is a member of the groups or not.

**Quick start guide videos are added across ODM** [Curators, Researchers].

??? example "Click here to expand"

    Each page contains a button that leads to a video explaining how to work with the page and the related main concepts. In addition, a welcome pop-up appears after the first login. The pop-up contains a link to the Quick Start Guide help page and all videos. Now users can get acquainted with the possibilities of ODM even faster and more conveniently on their own.

### Fixed issues

[Curators] The import_ODM_data.py script incorrectly handled unexpected server responses from URL provided with `--server`: server is down for any reason or a user mistake. Now the script returns an error with clear instructions.

<hr/>
